### PR TITLE
Don't crash with outdated share provider on update with the web updater

### LINKED ISSUE #34673: [Bug]: During Upgrade - PHP Fatal Error - Not compatible Declaration of RoomShareProvider::getSharesInFolder
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Tried to upgrade my latest Nextcloud, version 24 to the current 25.0.0 release.
After running ./occ upgrade it crashes with an PHP Fatal error:

> PHP Fatal error:  Declaration of OCA\Talk\Share\RoomShareProvider::getSharesInFolder($userId, OCP\Files\Folder $node, $reshares): array must be compatible with OCP\Share\IShareProvider::getSharesInFolder($userId, OCP\Files\Folder $node, $reshares, $shallow = true) 

Deactivated all incompatible apps before startet upgrade.

### Steps to reproduce

1. Upgrade from NC 24 to 25 in browser
3. run ./occ upgrade on console as recommended

### Expected behavior

No fatal error upgrade finishing slightly

### Installation method

Community Manual installation with Archive

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 7.4

### Web server

Apache (supported)

### Database engine version

SQlite

### Is this bug present after an update or on a fresh install?

Updated to a major version (ex. 22.2.3 to 23.0.1)

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "mydomain.tld"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "sqlite3",
        "version": "24.0.6.1",
        "overwrite.cli.url": "https:\/\/mydomain.tld\/",
        "installed": true,
        "loglevel": 3,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "remember_login_cookie_lifetime": 86400,
        "session_lifetime": 86400,
        "auth.bruteforce.protection.enabled": true,
        "versions_retention_obligation": "auto, 30",
        "default_language": "de",
        "default_locale": "de_DE",
        "knowledgebaseenabled": false,
        "htaccess.RewriteBase": "\/",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauth": 1,
        "mail_smtpauthtype": "LOGIN",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "overwritehost": "mydomain.tld",
        "maintenance": false,
        "theme": "",
        "data-fingerprint": "***REMOVED SENSITIVE VALUE***",
        "app_install_overwrite": [
            "keeweb",
            "bruteforcesettings",
            "ransomware_protection",
            "onlyoffice",
            "documentserver_community",
            "files_antivirus"
        ],
        "integrity.check.disabled": true,
        "default_phone_region": "DE",
        "3rdpartyroot": "\/var\/www\/nextcloud\/public_html\/"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - accessibility: 1.10.0
  - activity: 2.16.0
  - bruteforcesettings: 2.4.0
  - calendar: 3.5.0
  - circles: 24.0.1
  - cloud_federation_api: 1.7.0
  - contacts: 4.2.2
  - dav: 1.22.0
  - federatedfilesharing: 1.14.0
  - federation: 1.14.0
  - files: 1.19.0
  - files_accesscontrol: 1.14.1
  - files_antivirus: 3.3.1
  - files_pdfviewer: 2.5.0
  - files_rightclick: 1.3.0
  - files_sharing: 1.16.2
  - files_trashbin: 1.14.0
  - files_versions: 1.17.0
  - files_videoplayer: 1.13.0
  - firstrunwizard: 2.13.0
  - keeweb: 0.6.9
  - logreader: 2.9.0
  - lookup_server_connector: 1.12.0
  - nextcloud_announcements: 1.13.0
  - notifications: 2.12.1
  - oauth2: 1.12.0
  - password_policy: 1.14.0
  - photos: 1.6.0
  - privacy: 1.8.0
  - provisioning_api: 1.14.0
  - recommendations: 1.3.0
  - serverinfo: 1.14.0
  - settings: 1.6.0
  - sharebymail: 1.14.0
  - spreed: 14.0.5
  - tasks: 0.14.4
  - text: 3.5.1
  - theming: 1.15.0
  - twofactor_backupcodes: 1.13.0
  - updatenotification: 1.14.0
  - viewer: 1.8.0
  - workflowengine: 2.6.0
Disabled:
  - admin_audit
  - comments: 1.10.0
  - contactsinteraction: 1.2.0
  - dashboard: 7.1.0
  - encryption
  - files_external
  - ransomware_protection: 1.13.0
  - support: 1.1.0
  - survey_client: 1.6.0
  - systemtags: 1.10.0
  - user_ldap
  - user_status: 1.0.0
  - weather_status: 1.0.0
```

### Nextcloud Signing status

```shell
Integrity checker has been disabled. Integrity cannot be verified.
```

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #34673

### CODE REVIEW

File: lib/private/Share20/ProviderFactory.php [hunk 1]
```diff
@@ -257,8 +260,14 @@ protected function getRoomShareProvider() {
 			}
 
 			try {
-				$this->roomShareProvider = $this->serverContainer->query('\OCA\Talk\Share\RoomShareProvider');
-			} catch (\OCP\AppFramework\QueryException $e) {
+				$this->roomShareProvider = $this->serverContainer->get(RoomShareProvider::class);
+			} catch (\Throwable $e) {
+				if (!$e instanceof ContainerExceptionInterface) {
```
Comment (nickvergessen) [inline lib/private/Share20/ProviderFactory.php:265]: Not even sure we should do this or just log it all the time.

File: lib/private/Share20/ProviderFactory.php [hunk 2]
```diff
@@ -351,8 +361,17 @@ public function getAllProviders() {
 		}
 
 		foreach ($this->registeredShareProviders as $shareProvider) {
-			/** @var IShareProvider $instance */
-			$instance = $this->serverContainer->get($shareProvider);
+			try {
```
Comment (juliushaertl) [inline lib/private/Share20/ProviderFactory.php:364]: Would also be good in `getProvider` https://github.com/nextcloud/server/pull/34688/files#diff-5f34cda3371490875148d9accbf26ba847766f110f221c1eb7e8f2972b5da89eR282

Comment (nickvergessen) [reply lib/private/Share20/ProviderFactory.php:364]: Done

**Review by juliusknorr — APPROVED** [review]

**Review by szaimen — APPROVED** [review]
🐘

**Comment by blizzz** [comment]
Please run `composer run cs:fix` to format your code

**Comment by blizzz** [comment]
Catching Throwable in all cases is broad. Maybe ok here. Mixed feelings.

**Comment by nickvergessen** [comment]
> Please run composer run cs:fix to format your code

Fixed the unused import

**Pull request merged**